cpp/mrc/include/mrc/node/channel_holder.hpp

```diff
@@ -0,0 +1,1041 @@
+    // void add_lifetime(std::shared_ptr<EdgeHolder<T>> downstream)
+    // {
+    //     m_keep_alive.push_back(downstream);
+    // }
+
+  private:
+    bool m_is_connected{false};
+    std::vector<EdgeLifetime> m_connectors;
+    std::vector<EdgeLifetime> m_disconnectors;
+    std::vector<std::shared_ptr<EdgeTag>> m_linked_edges;
+
+    // Friend any type of edge handle to allow calling connect
+    template <typename>
+    friend class EdgeHandle;
+};
+
+template <typename T>
+class EdgeHandle : public virtual EdgeTag
+{
+  public:
+    // Friend the holder classes which are required to setup connections
+    template <typename>
+    friend class EdgeHolder;
+    template <typename, typename>
+    friend class MultiEdgeHolder;
+};
+
+struct EdgeTypePair
+{
+  public:
+    EdgeTypePair(const EdgeTypePair& other) = default;
+
+    std::type_index full_type() const
+    {
+        if (m_is_deferred)
+        {
+            throw std::runtime_error("Should not call full_type() for deferred types. Check is_deferred() first.");
+        }
+
+        return m_full_type.value();
+    }
+
+    std::type_index unwrapped_type() const
+    {
+        if (m_is_deferred)
+        {
+            throw std::runtime_error("Should not call unwrapped_type() for deferred types. Check is_deferred() first.");
+        }
+
+        return m_unwrapped_type.value();
+    }
+
+    bool is_deferred() const
+    {
+        return m_is_deferred;
+    }
+
+    bool operator==(const EdgeTypePair& other) const
+    {
+        return m_is_deferred == other.m_is_deferred && m_full_type == other.m_full_type &&
+               m_unwrapped_type == other.m_unwrapped_type;
+    }
+
+    template <typename T>
+    static EdgeTypePair create()
+    {
+        if constexpr (is_smart_ptr<T>::value)
+        {
+            return {typeid(T), typeid(typename T::element_type), false};
+        }
+        else
+        {
+            return {typeid(T), typeid(T), false};
+        }
+    }
+
+    static EdgeTypePair create_deferred()
+    {
+        return {std::nullopt, std::nullopt, true};
+    }
+
+  private:
+    EdgeTypePair(std::optional<std::type_index> full_type,
+                 std::optional<std::type_index> unwrapped_type,
+                 bool is_deferred) :
+      m_full_type(full_type),
+      m_unwrapped_type(unwrapped_type),
+      m_is_deferred(is_deferred)
+    {
+        CHECK((m_is_deferred && !m_full_type.has_value() && !m_unwrapped_type.has_value()) ||
+              (!m_is_deferred && m_full_type.has_value() && m_unwrapped_type.has_value()))
+            << "Inconsistent deferred setting with concrete types";
+    }
+
+    std::optional<std::type_index> m_full_type;       // Includes any wrappers like shared_ptr
+    std::optional<std::type_index> m_unwrapped_type;  // Excludes any wrappers like shared_ptr if they exist
+    bool m_is_deferred{false};                        // Whether or not this type is deferred or concrete
+};
+
+struct EdgeHandleObj
```

Comment:
It started off as a simple struct with 2 properties but got more complicated. Agreed it should be a class.

`Obj` is short for object and is just a way to differentiate the various class names. We have:

- `EdgeLifetime`
   - Holds a function object that gets called when destroyed. Can be armed or disarmed
- `EdgeTag`
   - Typeless base class for all edges. Holds onto `EdgeLifetime` objects. Maintains some state about the edge (whether or not its connected or disconnected)
- `EdgeHandle`
   - Typed `EdgeTag`
- `EdgeTypePair`
   - Type info about the edge including the wrapped and unwrapped types
- `EdgeHandleObj`
   - Object that holds onto a `EdgeTag` and `EdgeTypePair`.
   - Passed between `EdgeHolder` classes
   - Holds onto an edge to allow edge connection but prevents users from directly interacting with edges (i.e. we dont want users calling `await_write` directly on edge objects. They must be set to an `EdgeTag` 
- `EdgeHolder`
   - Primary class for holding onto edge objects
   - Maintains lifetime of edge objects
   - Base class for all nodes

Jotting these down for now. Open to ideas for different names

Reply:
I don't think 'Obj' is helpful, because, all of these types are Objects, so adding 'Obj' to some makes it confusing as to what Object is meant to convey. Some suggestions, based on description:

EdgeLifetime -> I'm not really opposed to this, although, you create EdgeLifetime objects in the code named 'connectors' which seems confusing to me.

EdgeTag -> Why not 'EdgeHandleTypeless', or 'EdgeHandleBase', something that conveys what its meant to be/do, and fits with most of our other naming schemes.

Off topic, but, Since its just holding an EdgeLifetime, is there a reason the edge destruction can't be handled in its destructor?

EdgeHandle -> Also seems fine.

EdgeHandleObj -> Really don't like the 'Obj' suffix, because it doesn't convey any information. Based on the description, something like 'EdgeDelegate' makes more sense to me. Because its not really a Handle, it contains a typeless handle+ some other information, and a callable interface.

EdgeTypePair -> Makes sense, describes exactly what it is.

EdgeHolder -> This seems fine too.

Reply:
Discussed in person. Resolving.